replace the leak check with universes, take 2

nikomatsakis opened the pull request:
This PR is an attempt to revive the "universe-based region check", which is an important step towards lazy normalization. Unlike before, we also modify the definition of `'empty` so that it is indexed by a universe. This sidesteps some of the surprising effects we saw before -- at the core, we no longer think that `exists<'a> { forall<'b> { 'b: 'a } }` is solveable. The new region lattice looks like this:

```
static ----------+-----...------+       (greatest)
|                |              |
early-bound and  |              |
free regions     |              |
|                |              |
scope regions    |              |
|                |              |
empty(root)   placeholder(U1)   |
|            /                  |
|           /         placeholder(Un)
empty(U1) --         /
|                   /
...                /
|                 /
empty(Un) --------                      (smallest)
```
This PR has three effects:

* It changes a fair number of error messages, I think for the better.
* It fixes a number of bugs. The old algorithm was too conservative and caused us to reject legal subtypings.
* It also causes two regressions (things that used to compile, but now do not).
    * `coherence-subtyping.rs` gets an additional error. This is expected.
    * `issue-57639.rs` regresses as before, for the reasons covered in #57639.

Both of the regressions stem from the same underlying property: without the leak check, the instantaneous "subtype" check is not able to tell whether higher-ranked subtyping will succeed or not. In both cases, we might be able to fix the problem by doing a 'leak-check like change' at some later point (e.g., as part of coherence).

This is a draft PR because:

* I didn't finish ripping out the leak-check completely.
* We might want to consider a crater run before landing this.
* We might want some kind of design meeting to cover the overall strategy.
* I just remembered I never finished 100% integrating this into the canonicalization code.
* I should also review what happens in NLL region checking -- it probably still has a notion of bottom (empty set).

r? @matthewjasper

Centril commented on src/librustc/infer/lexical_region_resolve/mod.rs:
```diff
@@ -83,6 +83,16 @@ pub enum RegionResolutionError<'tcx> {
     ),
 
+    /// Indicates a `'b: 'a` constraint where `'a` is in a universe that
+    /// cannot name the placeholder `'b`
```
```suggestion
    /// cannot name the placeholder `'b`.
```

Centril commented on src/librustc/infer/lexical_region_resolve/mod.rs:
```diff
@@ -402,14 +419,40 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
+            return true;
+        }
+
+        // If both a and b are free, consult the declared
```
```suggestion
        // If both `a` and `b` are free, consult the declared
```

Centril commented on src/librustc/infer/lexical_region_resolve/mod.rs:
```diff
@@ -402,14 +419,40 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
+        }
+
+        // For other cases, leverage the LUB code to find the LUB and
+        // check if it is equal to b.
```
```suggestion
        // check if it is equal to `b`.
```

Centril commented on src/librustc/infer/lexical_region_resolve/mod.rs:
```diff
@@ -437,6 +472,38 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
             }
 
+            (&ReStatic, _) | (_, &ReStatic) => {
+                // nothing lives longer than static
```
```suggestion
                // Nothing lives longer than `static`.
```

Centril commented on src/librustc/infer/lexical_region_resolve/mod.rs:
```diff
@@ -437,6 +472,38 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
+            (&ReEmpty(_), r @ ReEarlyBound(_)) | (r @ ReEarlyBound(_), &ReEmpty(_))
+            | (&ReEmpty(_), r @ ReFree(_)) | (r @ ReFree(_), &ReEmpty(_))
+            | (&ReEmpty(_), r @ ReScope(_)) | (r @ ReScope(_), &ReEmpty(_)) => {
+                // all empty regions are less than early-bound, free,
```
```suggestion
                // All empty regions are less than early-bound, free,
```

Centril commented on src/librustc/infer/lexical_region_resolve/mod.rs:
```diff
@@ -437,6 +472,38 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
+            | (&ReEmpty(_), r @ ReFree(_)) | (r @ ReFree(_), &ReEmpty(_))
+            | (&ReEmpty(_), r @ ReScope(_)) | (r @ ReScope(_), &ReEmpty(_)) => {
+                // all empty regions are less than early-bound, free,
+                // and scope regions
```
```suggestion
                // and scope regions.
```

Centril commented on src/librustc/infer/lexical_region_resolve/mod.rs:
```diff
@@ -437,6 +472,38 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
+            }
+
+            (&ReEmpty(a_ui), &ReEmpty(b_ui)) => {
+                // empty regions are ordered according to the universe
```
```suggestion
                // Empty regions are ordered according to the universe
```

Centril commented on src/librustc/infer/lexical_region_resolve/mod.rs:
```diff
@@ -437,6 +472,38 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
+
+            (&ReEmpty(a_ui), &ReEmpty(b_ui)) => {
+                // empty regions are ordered according to the universe
+                // they are associated with
```
```suggestion
                // they are associated with.
```

Centril commented on src/librustc/infer/outlives/free_region_map.rs:
```diff
@@ -23,11 +23,65 @@ impl<'tcx> FreeRegionMap<'tcx> {
+        }
+    }
+
+    /// Check whether `r_a <= r_b` is found in the relation
```
```suggestion
    /// Check whether `r_a <= r_b` is found in the relation.
```

Centril commented on src/librustc/infer/outlives/verify.rs:
```diff
@@ -64,7 +64,18 @@ impl<'cx, 'tcx> VerifyBoundCx<'cx, 'tcx> {
+            // being tested is `'empty`.
+            VerifyBound::IsEmpty
+        } else {
+            // If we can find any other bound R such that `T: R`, then
```
```suggestion
            // If we can find any other bound `R` such that `T: R`, then
```

Centril commented on src/librustc/ty/context.rs:
```diff
@@ -181,8 +181,13 @@ pub struct CommonTypes<'tcx> {
 
 pub struct CommonLifetimes<'tcx> {
-    pub re_empty: Region<'tcx>,
+    /// ReEmpty in the root universe
```
```suggestion
    /// `ReEmpty` in the root universe.
```

Centril commented on src/librustc/ty/context.rs:
```diff
@@ -181,8 +181,13 @@ pub struct CommonTypes<'tcx> {
+    /// ReEmpty in the root universe
+    pub re_root_empty: Region<'tcx>,
+
+    /// ReStatic
```
```suggestion
    /// `ReStatic`
```

Centril commented on src/librustc/ty/sty.rs:
```diff
@@ -1301,14 +1357,13 @@ pub enum RegionKind {
+    /// Empty lifetime is for data that is never accessed.  We tag the
+    /// empty lifetime with a universe -- the idea is that we don't
+    /// want `exists<'a> { forall<'b> { 'b: 'a } }` to be satisfiable.
+    /// Therefore, the `'empty` in a universe U is less than all
```
```suggestion
    /// Therefore, the `'empty` in a universe `U` is less than all
```

Centril commented on src/librustc/ty/sty.rs:
```diff
@@ -1301,14 +1357,13 @@ pub enum RegionKind {
+    /// empty lifetime with a universe -- the idea is that we don't
+    /// want `exists<'a> { forall<'b> { 'b: 'a } }` to be satisfiable.
+    /// Therefore, the `'empty` in a universe U is less than all
+    /// regions visible from U, but not less than regions not visible
```
```suggestion
    /// regions visible from `U`, but not less than regions not visible
```

Centril commented on src/librustc/ty/sty.rs:
```diff
@@ -1301,14 +1357,13 @@ pub enum RegionKind {
+    /// want `exists<'a> { forall<'b> { 'b: 'a } }` to be satisfiable.
+    /// Therefore, the `'empty` in a universe U is less than all
+    /// regions visible from U, but not less than regions not visible
+    /// from U.
```
```suggestion
    /// from `U`.
```

lqd commented:
When this PR is complete (particularly the "I didn't finish ripping out the leak-check completely." bullet point), it will also fix #59490.

matthewjasper commented:
> * I should also review what happens in NLL region checking -- it probably still has a notion of bottom (empty set).

It definitely considers regions that are not live anywhere and don't outlive any universal regions as `'empty`:

https://github.com/rust-lang/rust/blob/9c5a5c471aa9229b82bafa8ec18e55722d9ba8c3/src/librustc_mir/borrow_check/nll/region_infer/mod.rs#L1524-L1533

nikomatsakis commented:
@matthewjasper yeah, [see also this comment](https://github.com/rust-lang/rust/issues/57642#issuecomment-540012551)

nikomatsakis commented:
It occurs to me that updating NLL to match is not (presently) a strict requirement, but it would be something we have to do before we disable the AST-based regionck.

matthewjasper commented:
Can you bless the NLL compare mode output and check how this handles #57936 as well?

nikomatsakis commented:
@matthewjasper I will bless compare-mode output, yes. As far as #57936, I get two errors: one in the `indirect` function and one in the `direct` function. The `indirect` error appears to be new? It is as follows:

```
error[E0308]: mismatched types
  --> /home/nmatsakis/tmp/issue-57936.rs:19:5
   |
19 |     indirect::<fn(&())>(); // OK
   |     ^^^^^^^^^^^^^^^^^^^ one type is more general than the other
   |
   = note: expected type `X`
              found type `X`
```

This error is very bad -- I think we have an open issue on it, actually? I should fix that. The problem here is that the real cause has to do with the full trait ref, but for some reason we're not printing all the details.

nikomatsakis commented:
(I had not noticed #57936 at the time, or else I forgot about it, I'm not sure. :)

nikomatsakis commented:
Commit just pushed. Message:

--bless --compare-mode=nll -- but note that we get two errors

```
failures:
    [ui (nll)] ui/hrtb/hrtb-exists-forall-trait-contravariant.rs
    [ui (nll)] ui/hrtb/issue-46989.rs
```

Both are related to the NLL region checker not handling the `empty(ui)` correctly.

lqd commented:
> ```
> ...
>    = note: expected type `X`
>               found type `X`
> ```
> 
> This error is very bad -- I think we have an open issue on it, actually? 

We used to have #57362 for example, but it was fixed.

At the time we mentioned there might still be some cases missing from `try_report_placeholder_conflict`'s [big match](https://github.com/rust-lang/rust/blob/master/src/librustc/infer/error_reporting/nice_region_error/placeholder_error.rs#L20-L162) but didn't have cases exercising these possibly missing arms, maybe that is such a case ?

bors commented:
:umbrella: The latest upstream changes (presumably #64939) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
@lqd 

> We used to have #57362 for example, but it was fixed.

OK, well, I guess I broke it again. Thanks for the cite.

nikomatsakis commented on src/test/ui/issues/issue-57362-2.stderr:
```diff
@@ -1,13 +1,27 @@
-   = note: the following trait defines an item `make_g`, perhaps you need to implement it:
-           candidate #1: `X`
+   = note: expected type `X`
+              found type `X`
```
Yeah, I think before I land this I have to figure out what's going on there

lqd commented:
I looked a bit at the region errors managing to avoid the nice_region_errors mechanism, and left a couple notes in [this Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/leak.20check.20to.20universes.20take.202.20.2365232).

JohnCSimon commented:
Ping from triage.
@nikomatsakis can you please address the merge conflicts? 
Thank you!

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13253&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13253/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13253/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-11T14:13:15.1576502Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-11T14:13:15.1768153Z ##[command]git config gc.auto 0
2019-11-11T14:13:15.1822857Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-11T14:13:15.1855333Z ##[command]git config --get-all http.proxy
2019-11-11T14:13:15.2012362Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/65232/merge:refs/remotes/pull/65232/merge
---
2019-11-11T15:07:40.8233505Z .................................................................................................... 1400/9225
2019-11-11T15:07:46.5733984Z .................................................................................................... 1500/9225
2019-11-11T15:07:52.1154842Z .................................................................................................... 1600/9225
2019-11-11T15:08:00.7540114Z .................................................................................................... 1700/9225
2019-11-11T15:08:08.6262271Z ..i................................................................................................. 1800/9225
2019-11-11T15:08:14.8139133Z ......................................................................................iiiii......... 1900/9225
2019-11-11T15:08:34.3191944Z .................................................................................................... 2100/9225
2019-11-11T15:08:36.4470163Z .................................................................................................... 2200/9225
2019-11-11T15:08:38.7134379Z .................................................................................................... 2300/9225
2019-11-11T15:08:47.4838379Z .................................................................................................... 2400/9225
---
2019-11-11T15:11:27.5864845Z ...............................................................................i...............i.... 4700/9225
2019-11-11T15:11:34.3487281Z .................................................................................................... 4800/9225
2019-11-11T15:11:42.5444706Z .................................................................................................... 4900/9225
2019-11-11T15:11:47.1862820Z .................................................................................................... 5000/9225
2019-11-11T15:11:57.6723122Z ..................................................................................ii.ii...........i. 5100/9225
2019-11-11T15:12:05.7303677Z .................i.................................................................................. 5300/9225
2019-11-11T15:12:14.6746396Z .................................................................................................... 5400/9225
2019-11-11T15:12:21.0122275Z ................................................................i................................... 5500/9225
2019-11-11T15:12:27.6395903Z .................................................................................................... 5600/9225
2019-11-11T15:12:27.6395903Z .................................................................................................... 5600/9225
2019-11-11T15:12:35.8454811Z .................................................................................................... 5700/9225
2019-11-11T15:12:43.0986403Z .................................................ii...i..ii...........i............................. 5800/9225
2019-11-11T15:13:04.2365706Z .................................................................................................... 6000/9225
2019-11-11T15:13:11.8766614Z .................................................................................................... 6100/9225
2019-11-11T15:13:11.8766614Z .................................................................................................... 6100/9225
2019-11-11T15:13:16.7273441Z ....................................................................i..ii........................... 6200/9225
2019-11-11T15:13:43.2487225Z .................................................................................................... 6400/9225
2019-11-11T15:13:45.0658864Z ....................................i............................................................... 6500/9225
2019-11-11T15:13:47.0691202Z .................................................................................................... 6600/9225
2019-11-11T15:13:49.2123423Z ....................i............................................................................... 6700/9225
---
2019-11-11T15:18:13.0656749Z 
2019-11-11T15:18:13.0657390Z ---- [ui] ui/coherence/coherence-subtyping.rs#old stdout ----
2019-11-11T15:18:13.0657704Z diff of stderr:
2019-11-11T15:18:13.0657893Z 
2019-11-11T15:18:13.0658426Z 1 error[E0119]: conflicting implementations of trait `TheTrait` for type `for<'a, 'b> fn(&'a u8, &'b u8) -> &'a u8`:
2019-11-11T15:18:13.0658934Z -   --> $DIR/coherence-subtyping.rs:18:1
2019-11-11T15:18:13.0659917Z +   --> $DIR/coherence-subtyping.rs:16:1
2019-11-11T15:18:13.0660230Z 3    |
2019-11-11T15:18:13.0660739Z 4 LL | impl TheTrait for for<'a,'b> fn(&'a u8, &'b u8) -> &'a u8 {
2019-11-11T15:18:13.0661278Z 5    | --------------------------------------------------------- first implementation here
2019-11-11T15:18:13.0661737Z 
2019-11-11T15:18:13.0661956Z The actual stderr differed from the expected stderr.
2019-11-11T15:18:13.0661956Z The actual stderr differed from the expected stderr.
2019-11-11T15:18:13.0662494Z Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coherence/coherence-subtyping.old/coherence-subtyping.old.stderr
2019-11-11T15:18:13.0663407Z To update references, rerun the tests and pass the `--bless` flag
2019-11-11T15:18:13.0663881Z To only update this specific test, also pass `--test-args coherence/coherence-subtyping.rs`
2019-11-11T15:18:13.0664092Z 
2019-11-11T15:18:13.0664295Z error in revision `old`: 1 errors occurred comparing output.
2019-11-11T15:18:13.0664473Z status: exit code: 1
2019-11-11T15:18:13.0665356Z command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/coherence/coherence-subtyping.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "old" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coherence/coherence-subtyping.old" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coherence/coherence-subtyping.old/auxiliary" "-A" "unused"
2019-11-11T15:18:13.0666105Z ------------------------------------------
2019-11-11T15:18:13.0666325Z 
2019-11-11T15:18:13.0666710Z ------------------------------------------
2019-11-11T15:18:13.0666953Z stderr:
2019-11-11T15:18:13.0666953Z stderr:
2019-11-11T15:18:13.0667313Z ------------------------------------------
2019-11-11T15:18:13.0667777Z error[E0119]: conflicting implementations of trait `TheTrait` for type `for<'a, 'b> fn(&'a u8, &'b u8) -> &'a u8`:
2019-11-11T15:18:13.0668458Z    |
2019-11-11T15:18:13.0668458Z    |
2019-11-11T15:18:13.0668833Z LL | impl TheTrait for for<'a,'b> fn(&'a u8, &'b u8) -> &'a u8 {
2019-11-11T15:18:13.0670565Z    | --------------------------------------------------------- first implementation here
2019-11-11T15:18:13.0670904Z ...
2019-11-11T15:18:13.0671417Z LL | impl TheTrait for for<'a> fn(&'a u8, &'a u8) -> &'a u8 {
2019-11-11T15:18:13.0672007Z    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `for<'a, 'b> fn(&'a u8, &'b u8) -> &'a u8`
2019-11-11T15:18:13.0672287Z    |
2019-11-11T15:18:13.0672943Z    = note: this behavior recently changed as a result of a bug fix; see rust-lang/rust#56105 for details
2019-11-11T15:18:13.0673692Z error: aborting due to previous error
2019-11-11T15:18:13.0673859Z 
2019-11-11T15:18:13.0674248Z For more information about this error, try `rustc --explain E0119`.
2019-11-11T15:18:13.0674449Z 
2019-11-11T15:18:13.0674449Z 
2019-11-11T15:18:13.0674807Z ------------------------------------------
2019-11-11T15:18:13.0675027Z 
2019-11-11T15:18:13.0675175Z 
2019-11-11T15:18:13.0675564Z ---- [ui] ui/coherence/coherence-subtyping.rs#re stdout ----
2019-11-11T15:18:13.0675806Z diff of stderr:
2019-11-11T15:18:13.0675970Z 
2019-11-11T15:18:13.0676397Z 1 error[E0119]: conflicting implementations of trait `TheTrait` for type `for<'a, 'b> fn(&'a u8, &'b u8) -> &'a u8`:
2019-11-11T15:18:13.0676824Z -   --> $DIR/coherence-subtyping.rs:18:1
2019-11-11T15:18:13.0677377Z +   --> $DIR/coherence-subtyping.rs:16:1
2019-11-11T15:18:13.0677598Z 3    |
2019-11-11T15:18:13.0677999Z 4 LL | impl TheTrait for for<'a,'b> fn(&'a u8, &'b u8) -> &'a u8 {
2019-11-11T15:18:13.0678445Z 5    | --------------------------------------------------------- first implementation here
2019-11-11T15:18:13.0678826Z 
2019-11-11T15:18:13.0679347Z The actual stderr differed from the expected stderr.
2019-11-11T15:18:13.0679347Z The actual stderr differed from the expected stderr.
2019-11-11T15:18:13.0680060Z Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coherence/coherence-subtyping.re/coherence-subtyping.re.stderr
2019-11-11T15:18:13.0681278Z To update references, rerun the tests and pass the `--bless` flag
2019-11-11T15:18:13.0682016Z To only update this specific test, also pass `--test-args coherence/coherence-subtyping.rs`
2019-11-11T15:18:13.0682205Z 
2019-11-11T15:18:13.0682374Z error in revision `re`: 1 errors occurred comparing output.
2019-11-11T15:18:13.0682543Z status: exit code: 1
2019-11-11T15:18:13.0683550Z command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/coherence/coherence-subtyping.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "re" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coherence/coherence-subtyping.re" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coherence/coherence-subtyping.re/auxiliary" "-A" "unused"
2019-11-11T15:18:13.0684070Z ------------------------------------------
2019-11-11T15:18:13.0684223Z 
2019-11-11T15:18:13.0684527Z ------------------------------------------
2019-11-11T15:18:13.0684674Z stderr:
2019-11-11T15:18:13.0684674Z stderr:
2019-11-11T15:18:13.0684979Z ------------------------------------------
2019-11-11T15:18:13.0685448Z error[E0119]: conflicting implementations of trait `TheTrait` for type `for<'a, 'b> fn(&'a u8, &'b u8) -> &'a u8`:
2019-11-11T15:18:13.0685991Z    |
2019-11-11T15:18:13.0685991Z    |
2019-11-11T15:18:13.0686299Z LL | impl TheTrait for for<'a,'b> fn(&'a u8, &'b u8) -> &'a u8 {
2019-11-11T15:18:13.0686676Z    | --------------------------------------------------------- first implementation here
2019-11-11T15:18:13.0686821Z ...
2019-11-11T15:18:13.0687129Z LL | impl TheTrait for for<'a> fn(&'a u8, &'a u8) -> &'a u8 {
2019-11-11T15:18:13.0688951Z    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `for<'a, 'b> fn(&'a u8, &'b u8) -> &'a u8`
2019-11-11T15:18:13.0689883Z    |
2019-11-11T15:18:13.0690527Z    = note: this behavior recently changed as a result of a bug fix; see rust-lang/rust#56105 for details
2019-11-11T15:18:13.0690898Z error: aborting due to previous error
2019-11-11T15:18:13.0691021Z 
2019-11-11T15:18:13.0691452Z For more information about this error, try `rustc --explain E0119`.
2019-11-11T15:18:13.0691624Z 
---
2019-11-11T15:18:13.0703972Z thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:537:22
2019-11-11T15:18:13.0704213Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
2019-11-11T15:18:13.0707297Z 
2019-11-11T15:18:13.0707354Z 
2019-11-11T15:18:13.0715697Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-6.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "6.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
2019-11-11T15:18:13.0751514Z 
2019-11-11T15:18:13.0751897Z 
2019-11-11T15:18:13.0752216Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
2019-11-11T15:18:13.0752479Z Build completed unsuccessfully in 0:58:56
2019-11-11T15:18:13.0752479Z Build completed unsuccessfully in 0:58:56
2019-11-11T15:18:13.0768789Z == clock drift check ==
2019-11-11T15:18:13.0780266Z   local time: Mon Nov 11 15:18:13 UTC 2019
2019-11-11T15:18:13.3551207Z   network time: Mon, 11 Nov 2019 15:18:13 GMT
2019-11-11T15:18:13.3551391Z == end clock drift check ==
2019-11-11T15:18:14.2136442Z 
2019-11-11T15:18:14.2251436Z ##[error]Bash exited with code '1'.
2019-11-11T15:18:14.2286196Z ##[section]Starting: Checkout
2019-11-11T15:18:14.2287789Z ==============================================================================
2019-11-11T15:18:14.2287851Z Task         : Get sources
2019-11-11T15:18:14.2287893Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13305&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13305/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13305/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-11T22:02:27.2063660Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-11T22:02:27.2238856Z ##[command]git config gc.auto 0
2019-11-11T22:02:27.2316300Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-11T22:02:27.2368903Z ##[command]git config --get-all http.proxy
2019-11-11T22:02:27.2496935Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/65232/merge:refs/remotes/pull/65232/merge
---
2019-11-11T22:56:51.7647391Z .................................................................................................... 1400/9228
2019-11-11T22:56:57.6557540Z .................................................................................................... 1500/9228
2019-11-11T22:57:03.5510691Z .................................................................................................... 1600/9228
2019-11-11T22:57:12.4182452Z .................................................................................................... 1700/9228
2019-11-11T22:57:20.8519757Z ..i................................................................................................. 1800/9228
2019-11-11T22:57:27.3664908Z ......................................................................................iiiii......... 1900/9228
2019-11-11T22:57:47.6022956Z .................................................................................................... 2100/9228
2019-11-11T22:57:50.0103305Z .................................................................................................... 2200/9228
2019-11-11T22:57:52.4679351Z .................................................................................................... 2300/9228
2019-11-11T22:58:02.3423869Z .................................................................................................... 2400/9228
---
2019-11-11T23:00:56.2664681Z ..................................................................................i...............i. 4700/9228
2019-11-11T23:01:03.2843761Z .................................................................................................... 4800/9228
2019-11-11T23:01:12.3901829Z .................................................................................................... 4900/9228
2019-11-11T23:01:17.7407805Z .................................................................................................... 5000/9228
2019-11-11T23:01:29.2102733Z .....................................................................................ii.ii.......... 5100/9228
2019-11-11T23:01:33.0580251Z .i.................................................................................................. 5200/9228
2019-11-11T23:01:47.6396253Z .................................................................................................... 5400/9228
2019-11-11T23:01:54.6297852Z ...................................................................i................................ 5500/9228
2019-11-11T23:02:02.0087804Z .................................................................................................... 5600/9228
2019-11-11T23:02:09.9690698Z .................................................................................................... 5700/9228
2019-11-11T23:02:09.9690698Z .................................................................................................... 5700/9228
2019-11-11T23:02:19.0329938Z ....................................................ii...i..ii...........i.......................... 5800/9228
2019-11-11T23:02:41.6383994Z .................................................................................................... 6000/9228
2019-11-11T23:02:50.0969625Z .................................................................................................... 6100/9228
2019-11-11T23:02:50.0969625Z .................................................................................................... 6100/9228
2019-11-11T23:02:55.4023674Z .......................................................................i..ii........................ 6200/9228
2019-11-11T23:03:24.6834541Z .................................................................................................... 6400/9228
2019-11-11T23:03:26.7767033Z .......................................i............................................................ 6500/9228
2019-11-11T23:03:28.9522987Z .................................................................................................... 6600/9228
2019-11-11T23:03:31.2108286Z .......................i............................................................................ 6700/9228
---
2019-11-11T23:08:18.6817240Z 
2019-11-11T23:08:18.6818373Z ---- [ui] ui/coherence/coherence-subtyping.rs#old stdout ----
2019-11-11T23:08:18.6818600Z diff of stderr:
2019-11-11T23:08:18.6818730Z 
2019-11-11T23:08:18.6819194Z 1 error[E0119]: conflicting implementations of trait `TheTrait` for type `for<'a, 'b> fn(&'a u8, &'b u8) -> &'a u8`:
2019-11-11T23:08:18.6819571Z -   --> $DIR/coherence-subtyping.rs:18:1
2019-11-11T23:08:18.6819962Z +   --> $DIR/coherence-subtyping.rs:16:1
2019-11-11T23:08:18.6820155Z 3    |
2019-11-11T23:08:18.6820659Z 4 LL | impl TheTrait for for<'a,'b> fn(&'a u8, &'b u8) -> &'a u8 {
2019-11-11T23:08:18.6821041Z 5    | --------------------------------------------------------- first implementation here
2019-11-11T23:08:18.6821319Z 
2019-11-11T23:08:18.6821450Z The actual stderr differed from the expected stderr.
2019-11-11T23:08:18.6821450Z The actual stderr differed from the expected stderr.
2019-11-11T23:08:18.6821893Z Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coherence/coherence-subtyping.old/coherence-subtyping.old.stderr
2019-11-11T23:08:18.6822277Z To update references, rerun the tests and pass the `--bless` flag
2019-11-11T23:08:18.6822709Z To only update this specific test, also pass `--test-args coherence/coherence-subtyping.rs`
2019-11-11T23:08:18.6822889Z 
2019-11-11T23:08:18.6823030Z error in revision `old`: 1 errors occurred comparing output.
2019-11-11T23:08:18.6823163Z status: exit code: 1
2019-11-11T23:08:18.6824253Z command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/coherence/coherence-subtyping.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "old" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coherence/coherence-subtyping.old" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coherence/coherence-subtyping.old/auxiliary" "-A" "unused"
2019-11-11T23:08:18.6825233Z ------------------------------------------
2019-11-11T23:08:18.6825807Z 
2019-11-11T23:08:18.6826598Z ------------------------------------------
2019-11-11T23:08:18.6826820Z stderr:
2019-11-11T23:08:18.6826820Z stderr:
2019-11-11T23:08:18.6827197Z ------------------------------------------
2019-11-11T23:08:18.6827697Z error[E0119]: conflicting implementations of trait `TheTrait` for type `for<'a, 'b> fn(&'a u8, &'b u8) -> &'a u8`:
2019-11-11T23:08:18.6828683Z    |
2019-11-11T23:08:18.6828683Z    |
2019-11-11T23:08:18.6829371Z LL | impl TheTrait for for<'a,'b> fn(&'a u8, &'b u8) -> &'a u8 {
2019-11-11T23:08:18.6829771Z    | --------------------------------------------------------- first implementation here
2019-11-11T23:08:18.6829958Z ...
2019-11-11T23:08:18.6830318Z LL | impl TheTrait for for<'a> fn(&'a u8, &'a u8) -> &'a u8 {
2019-11-11T23:08:18.6830774Z    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `for<'a, 'b> fn(&'a u8, &'b u8) -> &'a u8`
2019-11-11T23:08:18.6830970Z    |
2019-11-11T23:08:18.6831519Z    = note: this behavior recently changed as a result of a bug fix; see rust-lang/rust#56105 for details
2019-11-11T23:08:18.6831824Z error: aborting due to previous error
2019-11-11T23:08:18.6831937Z 
2019-11-11T23:08:18.6833457Z For more information about this error, try `rustc --explain E0119`.
2019-11-11T23:08:18.6833654Z 
2019-11-11T23:08:18.6833654Z 
2019-11-11T23:08:18.6833977Z ------------------------------------------
2019-11-11T23:08:18.6834116Z 
2019-11-11T23:08:18.6834225Z 
2019-11-11T23:08:18.6835166Z ---- [ui] ui/coherence/coherence-subtyping.rs#re stdout ----
2019-11-11T23:08:18.6835393Z diff of stderr:
2019-11-11T23:08:18.6835523Z 
2019-11-11T23:08:18.6836001Z 1 error[E0119]: conflicting implementations of trait `TheTrait` for type `for<'a, 'b> fn(&'a u8, &'b u8) -> &'a u8`:
2019-11-11T23:08:18.6836401Z -   --> $DIR/coherence-subtyping.rs:18:1
2019-11-11T23:08:18.6836784Z +   --> $DIR/coherence-subtyping.rs:16:1
2019-11-11T23:08:18.6837190Z 3    |
2019-11-11T23:08:18.6837640Z 4 LL | impl TheTrait for for<'a,'b> fn(&'a u8, &'b u8) -> &'a u8 {
2019-11-11T23:08:18.6838107Z 5    | --------------------------------------------------------- first implementation here
2019-11-11T23:08:18.6838533Z 
2019-11-11T23:08:18.6838681Z The actual stderr differed from the expected stderr.
2019-11-11T23:08:18.6838681Z The actual stderr differed from the expected stderr.
2019-11-11T23:08:18.6839118Z Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coherence/coherence-subtyping.re/coherence-subtyping.re.stderr
2019-11-11T23:08:18.6839489Z To update references, rerun the tests and pass the `--bless` flag
2019-11-11T23:08:18.6840164Z To only update this specific test, also pass `--test-args coherence/coherence-subtyping.rs`
2019-11-11T23:08:18.6840492Z 
2019-11-11T23:08:18.6840643Z error in revision `re`: 1 errors occurred comparing output.
2019-11-11T23:08:18.6840810Z status: exit code: 1
2019-11-11T23:08:18.6841722Z command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/coherence/coherence-subtyping.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "re" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coherence/coherence-subtyping.re" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coherence/coherence-subtyping.re/auxiliary" "-A" "unused"
2019-11-11T23:08:18.6842478Z ------------------------------------------
2019-11-11T23:08:18.6843135Z 
2019-11-11T23:08:18.6843767Z ------------------------------------------
2019-11-11T23:08:18.6845074Z stderr:
2019-11-11T23:08:18.6845074Z stderr:
2019-11-11T23:08:18.6845646Z ------------------------------------------
2019-11-11T23:08:18.6846138Z error[E0119]: conflicting implementations of trait `TheTrait` for type `for<'a, 'b> fn(&'a u8, &'b u8) -> &'a u8`:
2019-11-11T23:08:18.6846520Z    |
2019-11-11T23:08:18.6846520Z    |
2019-11-11T23:08:18.6846766Z LL | impl TheTrait for for<'a,'b> fn(&'a u8, &'b u8) -> &'a u8 {
2019-11-11T23:08:18.6847033Z    | --------------------------------------------------------- first implementation here
2019-11-11T23:08:18.6847094Z ...
2019-11-11T23:08:18.6847355Z LL | impl TheTrait for for<'a> fn(&'a u8, &'a u8) -> &'a u8 {
2019-11-11T23:08:18.6847669Z    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `for<'a, 'b> fn(&'a u8, &'b u8) -> &'a u8`
2019-11-11T23:08:18.6847722Z    |
2019-11-11T23:08:18.6848022Z    = note: this behavior recently changed as a result of a bug fix; see rust-lang/rust#56105 for details
2019-11-11T23:08:18.6848276Z error: aborting due to previous error
2019-11-11T23:08:18.6848302Z 
2019-11-11T23:08:18.6848541Z For more information about this error, try `rustc --explain E0119`.
2019-11-11T23:08:18.6848572Z 
---
2019-11-11T23:08:18.6857779Z thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:537:22
2019-11-11T23:08:18.6857861Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
2019-11-11T23:08:18.6875997Z 
2019-11-11T23:08:18.6876125Z 
2019-11-11T23:08:18.6877963Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-6.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "6.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
2019-11-11T23:08:18.6878396Z 
2019-11-11T23:08:18.6878425Z 
2019-11-11T23:08:18.6917714Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
2019-11-11T23:08:18.6917811Z Build completed unsuccessfully in 0:59:36
2019-11-11T23:08:18.6917811Z Build completed unsuccessfully in 0:59:36
2019-11-11T23:08:18.6937667Z == clock drift check ==
2019-11-11T23:08:18.6953391Z   local time: Mon Nov 11 23:08:18 UTC 2019
2019-11-11T23:08:18.8458366Z   network time: Mon, 11 Nov 2019 23:08:18 GMT
2019-11-11T23:08:18.8462861Z == end clock drift check ==
2019-11-11T23:08:19.5991515Z 
2019-11-11T23:08:19.6102299Z ##[error]Bash exited with code '1'.
2019-11-11T23:08:19.6142956Z ##[section]Starting: Checkout
2019-11-11T23:08:19.6145278Z ==============================================================================
2019-11-11T23:08:19.6145336Z Task         : Get sources
2019-11-11T23:08:19.6145403Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

matthewjasper commented:
> * I just remembered I never finished 100% integrating this into the canonicalization code.

If this is referring to this, I guess this can be done in a follow up issue
https://github.com/rust-lang/rust/blob/94fe66d3df05a5f6329bae64e02dfe5ba09eb71a/src/librustc/infer/canonical/canonicalizer.rs#L202-L204

I think that this should be good to go with follow up issues raised and a crater run.

JohnCSimon commented:
Ping from triage:
@nikomatsakis Can you please address the failing build? 
cc: @matthewjasper @Centril 
Thanks.

bors commented:
:umbrella: The latest upstream changes (presumably #66578) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
@matthewjasper I have rebased this PR and I have converted it to issue warnings. I made them deny-by-default to start. We should probably do a crater run. Deny-by-default seemed reasonable since the last time we made this transition, we went straight to a hard error (see history in https://github.com/rust-lang/rust/issues/56105), owing to the clean crater run. That change was later backed out for independent reasons, but it didn't seem like there were a lot of people complaining in the interim. (At least based on the comments from the tracking issue.)

nikomatsakis commented:
I am however still running x.py test --bless...

nikomatsakis commented:
@bors try

bors commented:
:hourglass: Trying commit 69a72209aa8bfd202ac20499f3615432f9af4221 with merge a020ba7bcf99e5f54cfb3284fc331482915203dc...
<!-- homu: {"type":"TryBuildStarted","head_sha":"69a72209aa8bfd202ac20499f3615432f9af4221","merge_sha":"a020ba7bcf99e5f54cfb3284fc331482915203dc"} -->

nikomatsakis commented:
I probably won't be at keyboard when try build completes, but if somebody starts crater I'd be much obliged :)

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=19300)
Build commit: a020ba7bcf99e5f54cfb3284fc331482915203dc (`a020ba7bcf99e5f54cfb3284fc331482915203dc`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=19300"},"merge_sha":"a020ba7bcf99e5f54cfb3284fc331482915203dc"} -->

nikomatsakis commented:
@craterbot check

Centril commented:
Looks like the last regression is spurious, so only 2, which is pretty good. The only notable regression is `wasm-bindgen-0.2.58`. I think if we can fix upstream on nightly we should be able to land this without a warning period. cc @ashleygwilliams @fitzgen

nikomatsakis commented:
Based on these crater results, I think we should land this "as is" but with the FCW set to warn. We can then begin work on follow-up PRs that cleanup the support in NLL and which turn this into a hard error (or perhaps a deny-by-default lint).

Skimming the history, I saw these work items:

* extend support in NLL
* extend support in canonicalization, maybe (see [this comment](https://github.com/rust-lang/rust/pull/65232#issuecomment-553001315))

@matthewjasper also had a number of review comments that are now outdated. I'll have to rebase and see if they still apply.

@matthewjasper, do you want to review this "for real" post rebase? I forget if you r+'d in some form, I didn't notice any comment to that effect earlier.

matthewjasper commented:
I'll try to go through this one last time once you've rebased

nikomatsakis commented:
Also, to clarify for @ashleygwilliams and @fitzgen:

* Landing this PR will trigger warnings in wasm-bindgen
* I want to investigate exactly what's going on there
* The old code was accepting things that should never have been accepted
* This code is currently the **most conservative** it can be (i.e., it warns for all cases that may ultimately want to change)
* There is room to make it **less conservative** but that will require a bit more work and it would be useful to see what kinds of things start to get warnings
* So TL;DR I don't think any action is needed immediately for wasm-bindgen but obviously we should resolve that before we make this a hard error

fitzgen commented:
I'm not really sure what this PR is changing in a user-visible fashion -- is there some sort of summary I can look at somewhere to help understand what needs to change in `wasm-bindgen`?

The linked code that breaks in wasm-bindgen is our implementation of turning Rust closures into JS closures, FWIW. We have a `WasmClosure` trait that describes how to invoke the closure from across a raw wasm abi boundary. We implement it for a bunch of `dyn Fn(..)` and `dyn mut FnMut(...)` things inside a macro because we don't have a way to implement it once for all different function arities so things get pretty repetitive.

fitzgen commented:
+cc @alexcrichton

matthewjasper commented:
@fitzgen 
The issue that the following impls (and similar impls for `dyn FnMut` and of 4 or so other traits):

```rust
// Explicitly writing the bound lifetime.
impl<'a, 'b, A, R> IntoWasmAbi for &'a (dyn for<'x> Fn(&'x A) -> R + 'b)
where
    A: RefFromWasmAbi,
    R: ReturnWasmAbi
{ /* ... */ }
// From a macro:
impl<'a, 'b, A, R> IntoWasmAbi for &'a (dyn Fn(A) -> R + 'b)
    where A: FromWasmAbi,
        R: ReturnWasmAbi
{ /* ... */ }
```
Currently these are considered not to overlap because we don't consider it possible for `A` in the second impl to equal `&'x A` in the first.

The simplest way to work around this is to replace the first impl with separate impls for all of the actual types that implement `RefFromWasmAbi`, which would allow the compiler to use its knowledge of which types implement `FromWasmAbi` to ensure that there's no overlap.

@nikomatsakis Can you check that NLL compare mode works?

fitzgen commented:
> The simplest way to work around this is to replace the first impl with separate impls for all of the actual types that implement `RefFromWasmAbi`, which would allow the compiler to use its knowledge of which types implement `FromWasmAbi` to ensure that there's no overlap.

Unfortunately, every type that downstream users expose to JS via

```rust
#[wasm_bindgen]
pub struct MyTypeExposedToJs {
    // ...
}
```

implements `RefFromWasmAbi` so listing them all is impossible for us.

Is there another work around?

nikomatsakis commented:
> Currently these are considered not to overlap because we don't consider it possible for `A` in the second impl to equal `&'x A` in the first.

The question, I guess, is whether we *ought* to be warning in cases like these. 

Two considerations:

* We don't have variance-based subtyping for `dyn` types, so really I think it's true that there is no type that could be used for `A` to make it equal to `for<'x> &'x A`.
* For `fn` types, which do have variance for their argument types, the story could be different: if we have a `'empty` region (which I think is not fully settled?), then there would be overlap, but otherwise not.

Does this sound right? (This is partly what I meant by, we might ultimately tune the code to accept more, particularly if there is extant code that we really feel we have to continue accepting, and there is no easy workaround, which might be true here.)

Leave aside for a moment the way we would implement this distinction. But basically I think the question is something like this: equating the argument types yields some set of region constraints RC. Right now in the PR, we assume they are always satisfiable, but we could try to provide they are not satisfiable using *some* decision procedure.

alexcrichton commented:
Reading over this I don't know a ton about all the liftimes/checks/etc, but it sounds like this will add a warning to the compilation of the wasm-bindgen crate and that's pretty much it. If that's the case I think it's totally fine to go ahead and land this.

Also from reading this I don't think it's going to be easy/clear how we're going to fix the warning in wasm-bindgen (and let rustc transition to a hard error at some point), but it may be best to start a dedicated thread for that elsewhere?

nikomatsakis commented:
I agree that this doesn't block us from landing, but on the other hand I do think we have to resolve the question of how to migrate wasm bindgen. Currently I suspect the answer is that we'll want to alter the check and accept the code (and I believe we can, though I have to think a bit about how best to implement it).

nikomatsakis commented:
@matthewjasper `x.py test` and `x.py test --compare-mode=nll` both work for me :)

matthewjasper commented:
@bors r+

bors commented:
:pushpin: Commit 4b3c66d2c309a16d48c2b7f992a2038016a098d3 has been approved by `matthewjasper`

<!-- @bors r=matthewjasper 4b3c66d2c309a16d48c2b7f992a2038016a098d3 -->
<!-- homu: {"type":"Approved","sha":"4b3c66d2c309a16d48c2b7f992a2038016a098d3","approver":"matthewjasper"} -->

bors commented:
:hourglass: Testing commit 4b3c66d2c309a16d48c2b7f992a2038016a098d3 with merge 3ff7dbcff3cc38c0850f1ed6a409188c4bbcb36b...
<!-- homu: {"type":"BuildStarted","head_sha":"4b3c66d2c309a16d48c2b7f992a2038016a098d3","merge_sha":"3ff7dbcff3cc38c0850f1ed6a409188c4bbcb36b"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

mark-i-m commented:
@matthewjasper Perhaps retry?

matthewjasper commented:
@bors retry

bors commented:
:hourglass: Testing commit 4b3c66d2c309a16d48c2b7f992a2038016a098d3 with merge 92e5173d86cfea6987c917c74cb611fd330f959d...
<!-- homu: {"type":"BuildStarted","head_sha":"4b3c66d2c309a16d48c2b7f992a2038016a098d3","merge_sha":"92e5173d86cfea6987c917c74cb611fd330f959d"} -->

rust-highfive commented:
Your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=20376&view=logs&jobId=c9ea9f88-3fea-5145-c269-3bbafa17f0d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/20376/logs/1373), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/20376/logs/1373)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=20376)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=20376","builder_name":"checks-azure"} -->

lqd commented:
> The install of msys2 was NOT successful.

@bors retry

bors commented:
:hourglass: Testing commit 4b3c66d2c309a16d48c2b7f992a2038016a098d3 with merge 8498c5f5b02dbb4ed58a1eb4901b0b733342c35f...
<!-- homu: {"type":"BuildStarted","head_sha":"4b3c66d2c309a16d48c2b7f992a2038016a098d3","merge_sha":"8498c5f5b02dbb4ed58a1eb4901b0b733342c35f"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=20384)
Approved by: matthewjasper
Pushing 8498c5f5b02dbb4ed58a1eb4901b0b733342c35f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"matthewjasper","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=20384"},"merge_sha":"8498c5f5b02dbb4ed58a1eb4901b0b733342c35f"} -->

jonas-schievink commented:
This [regressed](https://perf.rust-lang.org/compare.html?start=a29424a2265411dda7d7446516ac5fd7499e2b55&end=8498c5f5b02dbb4ed58a1eb4901b0b733342c35f&stat=instructions:u) the `packed-simd` benchmark by ~5%, presumably due to the extra call to `traits::overlapping_impls`.

jonas-schievink commented:
I have opened a fix at https://github.com/rust-lang/rust/pull/69044
